<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Describe Glycosylation Structural Properties in a Site Specific Manner â€¢ glydet</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Describe Glycosylation Structural Properties in a Site Specific Manner">
<meta name="description" content="Calculate and analyze glycosylation derived traits for site-specific glycoproteomics data. Derived traits are summary features that aggregate individual glycan abundances into biologically meaningful groups (e.g., galactosylation, sialylation, fucosylation, branching). This package provides functions to compute well-established derived traits from the literature, and features a domain-specific language for defining custom derived traits tailored to specific research needs.">
<meta property="og:description" content="Calculate and analyze glycosylation derived traits for site-specific glycoproteomics data. Derived traits are summary features that aggregate individual glycan abundances into biologically meaningful groups (e.g., galactosylation, sialylation, fucosylation, branching). This package provides functions to compute well-established derived traits from the literature, and features a domain-specific language for defining custom derived traits tailored to specific research needs.">
<meta property="og:image" content="https://glycoverse.github.io/glydet/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">glydet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/glydet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/custom-traits.html">Defining Custom Traits</a></li>
    <li><a class="dropdown-item" href="articles/quantify-motifs.html">Quantifying Glycan Motifs</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="glydet-">glydet 
<a class="anchor" aria-label="anchor" href="#glydet-"></a>
</h1>
</div>
<!-- badges: start -->

<p>The goal of glydet is to describe glycosylation structural properties in a site specific manner. In the field of glycomics, this analytical approach is known as derived traits. glydet provides functions to calculate derived traits well-defined in literature, and implements a domain-specific language to define custom derived traits.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the latest release of glydet from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"glycoverse/glydet@*release"</span><span class="op">)</span></span></code></pre></div>
<p>Or install the development version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"glycoverse/glydet"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<ul>
<li>ğŸš€ Get started: <a href="https://glycoverse.github.io/glydet/articles/glydet.html">Here</a>
</li>
<li>ğŸ”§ Custom derived traits: <a href="https://glycoverse.github.io/glydet/articles/custom-traits.html">Here</a>
</li>
<li>ğŸ“š Reference: <a href="https://glycoverse.github.io/glydet/reference/index.html">Here</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="role-in-glycoverse">Role in <code>glycoverse</code>
<a class="anchor" aria-label="anchor" href="#role-in-glycoverse"></a>
</h2>
<p>glydet is a high-level package in the <code>glycoverse</code> ecosystem. It is designed to be used by glycomics or glycoproteomics researchers directly to calculate derived traits. It is built on top of many other packages in the <code>glycoverse</code> ecosystem, including <code>glyexp</code>, <code>glyrepr</code>, <code>glyparse</code>, and <code>glymotif</code>.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>First, letâ€™s load necessary packages and get the data ready.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">#&gt; Attaching package: 'glyclean'</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co">#&gt; The following object is masked from 'package:stats':</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#&gt;     aggregate</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#&gt; â”€â”€ Normalizing data â”€â”€</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt; â„¹ No QC samples found. Using default normalization method based on experiment type.</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt; â„¹ Experiment type is "glycoproteomics". Using `normalize_median()`.</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt; âœ” Normalization completed.</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co">#&gt; â”€â”€ Removing variables with too many missing values â”€â”€</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co">#&gt; â„¹ No QC samples found. Using all samples.</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co">#&gt; â„¹ Applying preset "discovery"...</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co">#&gt; â„¹ Total removed: 24 (0.56%) variables.</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">#&gt; âœ” Variable removal completed.</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="co">#&gt; â”€â”€ Imputing missing values â”€â”€</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">#&gt; â„¹ No QC samples found. Using default imputation method based on sample size.</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="co">#&gt; â„¹ Sample size &lt;= 30, using `impute_sample_min()`.</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="co">#&gt; âœ” Imputation completed.</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a><span class="co">#&gt; â”€â”€ Aggregating data â”€â”€</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="co">#&gt; â„¹ Aggregating to "gfs" level</span></span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="co">#&gt; âœ” Aggregation completed.</span></span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a><span class="co">#&gt; â”€â”€ Normalizing data again â”€â”€</span></span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a><span class="co">#&gt; â„¹ No QC samples found. Using default normalization method based on experiment type.</span></span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a><span class="co">#&gt; â„¹ Experiment type is "glycoproteomics". Using `normalize_median()`.</span></span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a><span class="co">#&gt; âœ” Normalization completed.</span></span>
<span id="cb3-36"><a href="#cb3-36" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-37"><a href="#cb3-37" tabindex="-1"></a><span class="co">#&gt; â”€â”€ Correcting batch effects â”€â”€</span></span>
<span id="cb3-38"><a href="#cb3-38" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-39"><a href="#cb3-39" tabindex="-1"></a><span class="co">#&gt; â„¹ Batch column  not found in sample_info. Skipping batch correction.</span></span>
<span id="cb3-40"><a href="#cb3-40" tabindex="-1"></a><span class="co">#&gt; âœ” Batch correction completed.</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ Glycoproteomics Experiment â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span>
<span><span class="co">#&gt; â„¹ Expression matrix: 12 samples, 3979 variables</span></span>
<span><span class="co">#&gt; â„¹ Sample information fields: group &lt;fct&gt;</span></span>
<span><span class="co">#&gt; â„¹ Variable information fields: protein &lt;chr&gt;, glycan_composition &lt;glyrpr_c&gt;, glycan_structure &lt;glyrpr_s&gt;, protein_site &lt;int&gt;, gene &lt;chr&gt;</span></span></code></pre></div>
<p>Now, letâ€™s calculate some derived traits!</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trait_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp</span><span class="op">)</span></span>
<span><span class="va">trait_exp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; â”€â”€ Traitproteomics Experiment â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span>
<span><span class="co">#&gt; â„¹ Expression matrix: 12 samples, 3864 variables</span></span>
<span><span class="co">#&gt; â„¹ Sample information fields: group &lt;fct&gt;</span></span>
<span><span class="co">#&gt; â„¹ Variable information fields: protein &lt;chr&gt;, protein_site &lt;int&gt;, trait &lt;chr&gt;, gene &lt;chr&gt;, explanation &lt;chr&gt;</span></span></code></pre></div>
<p>VoilÃ ! What you see is a brand new <code>experiment()</code> object with â€œtraitomicsâ€ type. Think of it as your original datasetâ€™s sophisticated cousin ğŸ­ â€” instead of tracking â€œquantification of each glycan on each glycosite in each sample,â€ it now contains â€œthe value of each derived trait on each glycosite in each sample.â€</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">get_var_info</span><span class="op">(</span><span class="va">trait_exp</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 3,864 Ã— 6</span></span>
<span><span class="co">#&gt;    variable protein protein_site trait gene  explanation                        </span></span>
<span><span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt;          &lt;int&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;                              </span></span>
<span><span class="co">#&gt;  1 V1       A6NJW9            49 TM    CD8B2 Proportion of high-mannose glycansâ€¦</span></span>
<span><span class="co">#&gt;  2 V2       A6NJW9            49 TH    CD8B2 Proportion of hybrid glycans amongâ€¦</span></span>
<span><span class="co">#&gt;  3 V3       A6NJW9            49 TC    CD8B2 Proportion of complex glycans amonâ€¦</span></span>
<span><span class="co">#&gt;  4 V4       A6NJW9            49 MM    CD8B2 Abundance-weighted mean of mannoseâ€¦</span></span>
<span><span class="co">#&gt;  5 V5       A6NJW9            49 CA2   CD8B2 Proportion of bi-antennary glycansâ€¦</span></span>
<span><span class="co">#&gt;  6 V6       A6NJW9            49 CA3   CD8B2 Proportion of tri-antennary glycanâ€¦</span></span>
<span><span class="co">#&gt;  7 V7       A6NJW9            49 CA4   CD8B2 Proportion of tetra-antennary glycâ€¦</span></span>
<span><span class="co">#&gt;  8 V8       A6NJW9            49 TF    CD8B2 Proportion of fucosylated glycans â€¦</span></span>
<span><span class="co">#&gt;  9 V9       A6NJW9            49 TFc   CD8B2 Proportion of core-fucosylated glyâ€¦</span></span>
<span><span class="co">#&gt; 10 V10      A6NJW9            49 TFa   CD8B2 Proportion of arm-fucosylated glycâ€¦</span></span>
<span><span class="co">#&gt; # â„¹ 3,854 more rows</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">get_expr_mat</span><span class="op">(</span><span class="va">trait_exp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;    C1 C2 C3 H1 H2</span></span>
<span><span class="co">#&gt; V1  0  0  0  0  0</span></span>
<span><span class="co">#&gt; V2  0  0  0  0  0</span></span>
<span><span class="co">#&gt; V3  1  1  1  1  1</span></span>
<span><span class="co">#&gt; V4 NA NA NA NA NA</span></span>
<span><span class="co">#&gt; V5  1  1  1  1  1</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
<li><a href="SUPPORT.html">Getting help</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing glydet</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Bin Fu <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-8567-2997" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://CRAN.R-project.org/package=glydet" class="external-link"><img src="https://www.r-pkg.org/badges/version/glydet" alt="CRAN status"></a></li>
<li><a href="https://github.com/glycoverse/glydet/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/glycoverse/glydet/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/glycoverse/glydet" class="external-link"><img src="https://codecov.io/gh/glycoverse/glydet/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
