<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Quantify Motifs in an Experiment — quantify_motifs • glydet</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quantify Motifs in an Experiment — quantify_motifs"><meta name="description" content="This function quantifies motifs from glycomic or glycoproteomic profiles.
For glycomics data, it calculates motif quantifications directly.
For glycoproteomics data, each glycosite is treated as a separate glycome,
and motif quantifications are calculated in a site-specific manner.
The function takes a glyexp::experiment() object and returns a new glyexp::experiment()
object with motif quantifications. Instead of containing quantifications of individual glycans
on each glycosite in each sample, the new experiment contains quantifications
of each motif on each glycosite in each sample (for glycoproteomics data) or
motif quantifications in each sample (for glycomics data)."><meta property="og:description" content="This function quantifies motifs from glycomic or glycoproteomic profiles.
For glycomics data, it calculates motif quantifications directly.
For glycoproteomics data, each glycosite is treated as a separate glycome,
and motif quantifications are calculated in a site-specific manner.
The function takes a glyexp::experiment() object and returns a new glyexp::experiment()
object with motif quantifications. Instead of containing quantifications of individual glycans
on each glycosite in each sample, the new experiment contains quantifications
of each motif on each glycosite in each sample (for glycoproteomics data) or
motif quantifications in each sample (for glycomics data)."><meta property="og:image" content="https://glycoverse.github.io/glydet/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">glydet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/glydet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/custom-traits.html">Defining Custom Traits</a></li>
    <li><a class="dropdown-item" href="../articles/quantify-motifs.html">Quantifying Glycan Motifs</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Quantify Motifs in an Experiment</h1>

      <div class="d-none name"><code>quantify_motifs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function quantifies motifs from glycomic or glycoproteomic profiles.
For glycomics data, it calculates motif quantifications directly.
For glycoproteomics data, each glycosite is treated as a separate glycome,
and motif quantifications are calculated in a site-specific manner.</p>
<p>The function takes a <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> object and returns a new <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code>
object with motif quantifications. Instead of containing quantifications of individual glycans
on each glycosite in each sample, the new experiment contains quantifications
of each motif on each glycosite in each sample (for glycoproteomics data) or
motif quantifications in each sample (for glycomics data).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quantify_motifs</span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span>, method <span class="op">=</span> <span class="st">"relative"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-exp">exp<a class="anchor" aria-label="anchor" href="#arg-exp"></a></dt>
<dd><p>A <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> object. Before using this function,
you should preprocess the data using the <code>glyclean</code> package.
For glycoproteomics data, the data should be aggregated to the
"gfs" (glycoforms with structures) level using <code><a href="https://glycoverse.github.io/glyclean/reference/aggregate.html" class="external-link">glyclean::aggregate()</a></code>.
Also, please make sure that the <code>glycan_structure</code> column is present in the <code>var_info</code> table,
as not all glycoproteomics identification softwares provide this information.
"glycan_structure" can be a <code><a href="https://glycoverse.github.io/glyrepr/reference/glycan_structure.html" class="external-link">glyrepr::glycan_structure()</a></code> vector,
or a character vector of glycan structure strings supported by <code><a href="https://glycoverse.github.io/glyparse/reference/auto_parse.html" class="external-link">glyparse::auto_parse()</a></code>.</p></dd>


<dt id="arg-motifs">motifs<a class="anchor" aria-label="anchor" href="#arg-motifs"></a></dt>
<dd><p>A character vector of motif names, glycan structure strings,
or a 'glyrepr_structure' object.
For glycan structure strings, all formats supported by <code><a href="https://glycoverse.github.io/glyparse/reference/auto_parse.html" class="external-link">glyparse::auto_parse()</a></code> are accepted,
including IUPAC-condensed, WURCS, GlycoCT, and others.
If the vector is named, the names will be used as motif names.
Otherwise, IUPAC-condensed structure strings will be used as motif names.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string specifying the quantification method.
Must be either "absolute" or "relative". Default is "relative".
See "Relative and Absolute Motif Quantification" section for details.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A new <code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> object containing motif quantifications.
Instead of containing quantifications of individual glycans on each glycosite in each sample,
the new experiment contains quantifications of each motif on each glycosite in each sample
(for glycoproteomics data) or motif quantifications in each sample (for glycomics data).</p>
<p>For glycoproteomics data, with additional columns:</p><ul><li><p><code>protein</code>: protein ID</p></li>
<li><p><code>protein_site</code>: the glycosite position on the protein</p></li>
</ul><p>Other columns in the <code>var_info</code> table (e.g., <code>gene</code>) are retained if they have a "many-to-one"
relationship with glycosites (unique combinations of <code>protein</code>, <code>protein_site</code>).
That is, each glycosite cannot have multiple values for these columns.
<code>gene</code> is a common example, as a glycosite can only be associated with one gene.
Glycan descriptions are not such columns, as a glycosite can have multiple glycans,
thus having multiple descriptions.
Columns that do not have this relationship with glycosites will be dropped.
Don't worry if you cannot understand this logic—
just know that this function will do its best to preserve useful information.</p>
<p>The <code>sample_info</code> and <code>meta_data</code> tables are not modified,
except that the <code>exp_type</code> field in <code>meta_data</code> is set to "traitomics" for glycomics data
and "traitproteomics" for glycoproteomics data.</p>
    </div>
    <div class="section level2">
    <h2 id="relative-and-absolute-motif-quantification">Relative and Absolute Motif Quantification<a class="anchor" aria-label="anchor" href="#relative-and-absolute-motif-quantification"></a></h2>
    <p>Motif quantification can be performed in two ways: absolute and relative.
Do not confuse this with the absolute and relative quantification of glycans or glycopeptides.</p>
<p>In an omics context, absolute quantification means we can determine the actual concentration
(e.g., in mg/L) or counts (e.g., mRNA copy numbers),
while relative quantification means we can only compare the abundance of the same
molecule between different samples,
but the absolute values themselves are not meaningful.
Label-free quantification is a relative quantification method.</p>
<p>In the context of motif quantification,
absolute and relative refer to whether we normalize the motif quantifications
by the total abundance of the glycome (or the mini-glycomes of individual glycosites).</p>
<p>For example, let's say we have a glycomics dataset with two samples, A and B.
In each sample, three glycans (G1, G2, G3) are present,
with the following abundances:</p><ul><li><p>Sample A: G1 = 10, G2 = 20, G3 = 30</p></li>
<li><p>Sample B: G1 = 20, G2 = 40, G3 = 60</p></li>
</ul><p>For simplicity, let's say the motif we want to quantify happens to appear once in all glycans.</p>
<p>For absolute motif quantification, we simply sum the abundances of the motif across all glycans:</p><ul><li><p>Sample A: 10 × 1 + 20 × 1 + 30 × 1 = 60  (× 1 because the motif appears once in each glycan)</p></li>
<li><p>Sample B: 20 × 1 + 40 × 1 + 60 × 1 = 120</p></li>
</ul><p>The results are clearly different.</p>
<p>However, if we quantify the motif in a relative way:</p><ul><li><p>Sample A: (10 × 1 + 20 × 1 + 30 × 1) / (10 + 20 + 30) = 60 / 60 = 1</p></li>
<li><p>Sample B: (20 × 1 + 40 × 1 + 60 × 1) / (20 + 40 + 60) = 120 / 120 = 1</p></li>
</ul><p>The results are identical!</p>
<p>The absolute motif quantification answers the question "how many motifs are there in the sample?",
while the relative motif quantification answers the question
"if I take out one glycan molecule, how many motifs are on it in average?"</p>
<p>So which method should you use?
It depends on both your data type and research objectives.
Here are some general guidelines:</p><ul><li><p>For glycomics data, you should typically use relative motif quantification,
because glycomics data is inherently compositional (search "compositional data" for more details).</p></li>
<li><p>For glycoproteomics data, the choice depends on your research question.
If you want to compare observed motif abundance across samples, use absolute motif quantification.
Note that many factors can cause one sample to have higher values than another,
such as upregulation of enzymes responsible for the motif,
or higher overall glycosylation site occupancy.
If you want to understand the underlying regulatory mechanisms,
use relative motif quantification to correct for differences in site occupancy.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="relationship-with-derived-traits">Relationship with Derived Traits<a class="anchor" aria-label="anchor" href="#relationship-with-derived-traits"></a></h2>
    <p>Motif quantification is a special type of derived trait.
It is simply a weighted sum (<code><a href="wsum.html">wsum()</a></code>) or weighted mean (<code><a href="wmean.html">wmean()</a></code>) of the motif counts,
for absolute and relative motif quantification, respectively.
You can perform motif quantification manually using <code><a href="derive_traits.html">derive_traits()</a></code>.</p>
<p>Let's perform absolute motif quantification manually using <code><a href="derive_traits.html">derive_traits()</a></code>
to better understand the process (we will use custom column meta-properties here):</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="co"># Add the meta-properties to the variable information tibble</span></span>
<span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  nLx <span class="op">=</span> <span class="st">"Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>,  <span class="co"># Lewis x antigen</span></span>
<span>  nSLx <span class="op">=</span> <span class="st">"NeuAc(??-?)Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>  <span class="co"># Sialyl Lewis x antigen</span></span>
<span><span class="op">)</span></span>
<span><span class="va">exp_with_mps</span> <span class="op">&lt;-</span> <span class="fu">glymotif</span><span class="fu">::</span><span class="fu"><a href="https://glycoverse.github.io/glymotif/reference/add_motifs_int.html" class="external-link">add_motifs_int</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define the traits</span></span>
<span><span class="va">trait_fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Lx <span class="op">=</span> <span class="fu"><a href="../reference/wsum.html">wsum</a></span><span class="op">(</span><span class="va">nLx</span><span class="op">)</span>, SLx <span class="op">=</span> <span class="fu"><a href="../reference/wsum.html">wsum</a></span><span class="op">(</span><span class="va">nSLx</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate the traits</span></span>
<span><span class="fu"><a href="../reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp_with_mps</span>, trait_fns <span class="op">=</span> <span class="va">trait_fns</span>, mp_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nLx"</span>, <span class="st">"nSLx"</span><span class="op">)</span><span class="op">)</span></span></code></pre><p></p></div>
<p>The code snippet above is equivalent to:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="co"># Quantify the motifs</span></span>
<span><span class="fu"><a href="../reference/quantify_motifs.html">quantify_motifs</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span>, method <span class="op">=</span> <span class="st">"absolute"</span><span class="op">)</span></span></code></pre><p></p></div>
<p>In fact, this is essentially the implementation of the <code>quantify_motifs()</code> function,
except the actual implementation is more robust and user-friendly.</p>
<p>For relative motif quantification, simply replace <code><a href="wsum.html">wsum()</a></code> with <code><a href="wmean.html">wmean()</a></code>,
and everything else remains the same.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="derive_traits.html">derive_traits()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyexp/" class="external-link">glyexp</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyclean/" class="external-link">glyclean</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="va">real_experiment</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://glycoverse.github.io/glyclean/reference/auto_clean.html" class="external-link">auto_clean</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/slice_obs.html" class="external-link">slice_head_var</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Normalizing data (Median)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Normalizing data (Median) <span style="color: #B2B2B2;">[8ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Removing variables with &gt;50% missing values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Removing variables with &gt;50% missing values <span style="color: #B2B2B2;">[17ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Imputing missing values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Sample size &lt;= 30, using sample minimum imputation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Imputing missing values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Imputing missing values <span style="color: #B2B2B2;">[18ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Aggregating data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Aggregating data <span style="color: #B2B2B2;">[1.5s]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Normalizing data again</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Normalizing data again <span style="color: #B2B2B2;">[79ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  nLx <span class="op">=</span> <span class="st">"Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>,  <span class="co"># Lewis x antigen</span></span></span>
<span class="r-in"><span>  nSLx <span class="op">=</span> <span class="st">"NeuAc(??-?)Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>  <span class="co"># Sialyl Lewis x antigen</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">quantify_motifs</span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Traitproteomics Experiment</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Expression matrix: <span style="color: #0000BB;">12</span> samples, <span style="color: #0000BB;">8</span> variables</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Sample information fields: <span style="color: #00BB00;">group</span> <span style="color: #0000BB;">&lt;fct&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Variable information fields: <span style="color: #00BB00;">protein</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">protein_site</span> <span style="color: #0000BB;">&lt;int&gt;</span>, <span style="color: #00BB00;">motif</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">gene</span> <span style="color: #0000BB;">&lt;chr&gt;</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

