<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Defining Custom Traits • glydet</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Defining Custom Traits">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">glydet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/glydet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/custom-traits.html">Defining Custom Traits</a></li>
    <li><a class="dropdown-item" href="../articles/quantify-motifs.html">Quantifying Glycan Motifs</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Defining Custom Traits</h1>
            
      

      <div class="d-none name"><code>custom-traits.Rmd</code></div>
    </div>

    
    
<p>This vignette provides a comprehensive guide for defining custom
derived traits using <code>glydet</code>. Before proceeding, please
ensure you have read the “Get Started with glydet” vignette and are
familiar with the fundamental concepts of derived traits and
meta-properties.</p>
<p><strong>Note about new features:</strong></p>
<p>We introduced an experimental function <code><a href="../reference/make_trait.html">make_trait()</a></code>
since glydet 0.7.0. This function allows you to create a trait from
natural language. Check out the last section of this vignette for
details.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glydet/">glydet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyexp/" class="external-link">glyexp</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyclean/" class="external-link">glyclean</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyrepr/" class="external-link">glyrepr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://glycoverse.github.io/glyclean/reference/auto_clean.html" class="external-link">auto_clean</a></span><span class="op">(</span><span class="va">real_experiment</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="custom-traits">Custom Traits<a class="anchor" aria-label="anchor" href="#custom-traits"></a>
</h2>
<p><code>glydet</code> provides three trait factory functions for
creating custom derived traits:</p>
<ul>
<li>
<code><a href="../reference/prop.html">prop()</a></code> for calculating the abundance proportion of a
specific glycan subset</li>
<li>
<code><a href="../reference/ratio.html">ratio()</a></code> for computing the abundance ratio between two
glycan subsets</li>
<li>
<code><a href="../reference/wmean.html">wmean()</a></code> for calculating the weighted mean of a
quantitative property, weighted by glycan abundances</li>
</ul>
<p>All derived traits in <code>glydet</code> are constructed using these
three factory functions. The definitions of all built-in traits
demonstrate their usage:</p>
<ul>
<li>
<code>TM</code>: <code>prop(Tp == "highmannose")</code>
</li>
<li>
<code>TH</code>: <code>prop(Tp == "hybrid")</code>
</li>
<li>
<code>TC</code>: <code>prop(Tp == "complex")</code>
</li>
<li>
<code>MM</code>:
<code>wmean(nM, within = (Tp == "highmannose"))</code>
</li>
<li>
<code>CA2</code>:
<code>prop(nA == 2, within = (Tp == "complex"))</code>
</li>
<li>
<code>CA3</code>:
<code>prop(nA == 3, within = (Tp == "complex"))</code>
</li>
<li>
<code>CA4</code>:
<code>prop(nA == 4, within = (Tp == "complex"))</code>
</li>
<li>
<code>TF</code>: <code>prop(nF &gt; 0)</code>
</li>
<li>
<code>TFc</code>: <code>prop(nFc &gt; 0)</code>
</li>
<li>
<code>TFa</code>: <code>prop(nFa &gt; 0)</code>
</li>
<li>
<code>TB</code>: <code>prop(B)</code>
</li>
<li>
<code>GS</code>: <code>wmean(nS / nG)</code>
</li>
<li>
<code>AG</code>: <code>wmean(nG / nA)</code>
</li>
<li>
<code>TS</code>: <code>prop(nS &gt; 0)</code>
</li>
</ul>
<p>These definitions utilize meta-properties as building blocks. For
instance, <code>T</code> represents glycan type, <code>nM</code> denotes
the number of mannose residues, and so forth. To retrieve all available
built-in meta-properties, use <code>names(all_mp_fns())</code>.</p>
<p>The complete list of built-in meta-properties is provided below:</p>
<table class="table">
<colgroup>
<col width="17%">
<col width="28%">
<col width="17%">
<col width="37%">
</colgroup>
<thead><tr class="header">
<th>Name</th>
<th>Function</th>
<th>Type</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>T</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_glycan_type()</a></code></td>
<td>factor</td>
<td>Type of the glycan, either “complex”, “hybrid”, “highmannose”, or
“pausimannose”</td>
</tr>
<tr class="even">
<td><code>B</code></td>
<td><code><a href="../reference/n_glycan_type.html">has_bisecting()</a></code></td>
<td>logical</td>
<td>Whether the glycan has a bisecting GlcNAc</td>
</tr>
<tr class="odd">
<td><code>nA</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_antennae()</a></code></td>
<td>integer</td>
<td>Number of antennae</td>
</tr>
<tr class="even">
<td><code>nF</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_fuc()</a></code></td>
<td>integer</td>
<td>Number of fucoses</td>
</tr>
<tr class="odd">
<td><code>nFc</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_core_fuc()</a></code></td>
<td>integer</td>
<td>Number of core fucoses</td>
</tr>
<tr class="even">
<td><code>nFa</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_arm_fuc()</a></code></td>
<td>integer</td>
<td>Number of arm fucoses</td>
</tr>
<tr class="odd">
<td><code>nG</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_gal()</a></code></td>
<td>integer</td>
<td>Number of galactoses</td>
</tr>
<tr class="even">
<td><code>nGt</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_terminal_gal()</a></code></td>
<td>integer</td>
<td>Number of terminal galactoses</td>
</tr>
<tr class="odd">
<td><code>nS</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_sia()</a></code></td>
<td>integer</td>
<td>Number of sialic acids</td>
</tr>
<tr class="even">
<td><code>nM</code></td>
<td><code><a href="../reference/n_glycan_type.html">n_man()</a></code></td>
<td>integer</td>
<td>Number of mannoses</td>
</tr>
</tbody>
</table>
<p>The following sections provide detailed explanations of each factory
function.</p>
<div class="section level3">
<h3 id="trait-factories">Trait Factories<a class="anchor" aria-label="anchor" href="#trait-factories"></a>
</h3>
<div class="section level4">
<h4 id="prop">
<code>prop()</code><a class="anchor" aria-label="anchor" href="#prop"></a>
</h4>
<p><code><a href="../reference/prop.html">prop()</a></code> creates proportion traits, which represent the
most common type of derived traits. These traits calculate the relative
abundance of a specific glycan subset within a defined population.
Examples include the proportion of core-fucosylated glycans within all
glycans, or the proportion of tetra-antennary glycans within the complex
glycan subset.</p>
<p><code><a href="../reference/prop.html">prop()</a></code> accepts an expression that evaluates to a logical
vector. Both built-in and custom meta-properties (covered later) can be
referenced within the expression.</p>
<p>For example, the proportion of core-fucosylated glycans within all
glycans is defined as:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nFc</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Since <code>nFc</code> is an integer meta-property representing the
number of core fucoses, the expression <code>nFc &gt; 0</code> evaluates
to a logical vector, creating a valid trait definition.</p>
<p>Consider this simpler example:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">B</span><span class="op">)</span></span></code></pre></div>
<p>This demonstrates a straightforward trait definition. Since
<code>B</code> (bisecting GlcNAc presence) is already a logical
meta-property, <code>prop(B)</code> constitutes a valid trait
definition.</p>
<p>A more complex example demonstrates compound logical operations:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nS</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">nFa</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>This trait calculates the proportion of glycans containing both
sialic acid and arm fucose. Both <code>nS &gt; 0</code> and
<code>nFa &gt; 0</code> represent logical expressions, combined using
the logical AND operator (<code>&amp;</code>). Any R logical operator
(including <code>|</code>, <code>!</code>, etc.) can be utilized within
these expressions.</p>
<p>Arithmetic calculations can also be incorporated within expressions.
For example, the <code>CF</code> trait is defined as:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nF</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>We know that <code>nFc + nFa</code> is equivalent to <code>nF</code>,
so the above definition is equivalent to:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="op">(</span><span class="va">nFc</span> <span class="op">+</span> <span class="va">nFa</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>As a general principle, any R expression that evaluates to a logical
vector is valid for use with <code><a href="../reference/prop.html">prop()</a></code>.</p>
<p>The traits described above calculate proportions relative to the
entire glycan population. To customize the denominator (reference
population), the <code>within</code> parameter can be employed. For
example, calculating the proportion of bi-antennary glycans within the
<strong>complex</strong> glycan subset (<code>CA2</code>) requires this
approach.</p>
<p>The <code>within</code> parameter accepts an expression that
evaluates to a logical vector, following the same syntax as the primary
parameter.</p>
<p>The <code>CA2</code> trait is defined as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">2</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that parentheses around <code>Tp == "complex"</code> are
optional. Consequently, this alternative definition is equally
valid:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">2</span>, within <span class="op">=</span> <span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span><span class="op">)</span></span></code></pre></div>
<p>Combining these two parameters enables the creation of sophisticated
trait definitions. For instance, the proportion of sialylated glycans
within the subset of core-fucosylated tetra-antennary glycans is defined
as:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nS</span> <span class="op">&gt;</span> <span class="fl">0</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">nFc</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">nA</span> <span class="op">==</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Having mastered the usage of <code><a href="../reference/prop.html">prop()</a></code>, examine the
definitions of built-in proportion traits to reinforce your
understanding.</p>
<p>Practice exercises:</p>
<ol style="list-style-type: decimal">
<li>Define a trait calculating the proportion of glycans containing
terminal galactose.</li>
<li>Define a trait calculating the proportion of glycans with terminal
galactose but no sialic acid.</li>
<li>Define a trait calculating the proportion of glycans with exactly
two antennae and no bisecting GlcNAc.</li>
<li>Define a trait calculating the proportion of glycans with bisecting
GlcNAc within the bi-antennary glycan subset.</li>
</ol>
<p>Solutions are provided at the end of this vignette.</p>
</div>
<div class="section level4">
<h4 id="ratio">
<code>ratio()</code><a class="anchor" aria-label="anchor" href="#ratio"></a>
</h4>
<p><code><a href="../reference/ratio.html">ratio()</a></code> creates ratio traits that represent the quotient
of total abundances between two glycan groups. Examples include the
ratio of complex to hybrid glycans, or the ratio of bisecting to
non-bisecting glycans.</p>
<p><code><a href="../reference/ratio.html">ratio()</a></code> accepts two expressions that evaluate to logical
vectors, similar to <code><a href="../reference/prop.html">prop()</a></code>.</p>
<p>For example, the ratio of complex to hybrid glycans is defined
as:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ratio.html">ratio</a></span><span class="op">(</span><span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span>, <span class="va">Tp</span> <span class="op">==</span> <span class="st">"hybrid"</span><span class="op">)</span></span></code></pre></div>
<p>The first expression defines the numerator, while the second
expression defines the denominator.</p>
<p>The <code>within</code> parameter can be employed to apply
restrictions to both numerator and denominator. For example, the ratio
of bisecting to non-bisecting glycans within the bi-antennary subset is
defined as:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ratio.html">ratio</a></span><span class="op">(</span><span class="va">B</span>, <span class="op">!</span><span class="va">B</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This represents syntactic sugar for the more verbose:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ratio.html">ratio</a></span><span class="op">(</span><span class="va">B</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span>, <span class="op">(</span><span class="op">!</span><span class="va">B</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>within</code> parameter provides clearer semantics and
reduces redundancy.</p>
<p>Note that <code><a href="../reference/prop.html">prop()</a></code> represents a special case of
<code><a href="../reference/ratio.html">ratio()</a></code>, where <code>prop(cond, within)</code> is
mathematically equivalent to
<code>ratio(cond &amp; within, within)</code>. The following two
definitions are functionally identical:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># using prop()</span></span>
<span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nFc</span> <span class="op">&gt;</span> <span class="fl">0</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using ratio()</span></span>
<span><span class="fu"><a href="../reference/ratio.html">ratio</a></span><span class="op">(</span><span class="va">nFc</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span>, <span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span><span class="op">)</span></span></code></pre></div>
<p>However, <code><a href="../reference/prop.html">prop()</a></code> is recommended when appropriate for
enhanced readability, as it provides more intuitive semantics that align
with natural language. The <code><a href="../reference/prop.html">prop()</a></code> example above can be
interpreted as “the proportion of core-fucosylated glycans within all
complex glycans,” whereas the <code><a href="../reference/ratio.html">ratio()</a></code> version requires
additional cognitive processing.</p>
<p>Practice exercises:</p>
<ol style="list-style-type: decimal">
<li>Define a trait calculating the ratio of core-fucosylated to
non-core-fucosylated glycans.</li>
<li>Define a trait calculating the ratio of tetra-antennary to
tri-antennary glycans within the complex glycan subset.</li>
<li>Redefine the <code>CA2</code> trait using <code><a href="../reference/ratio.html">ratio()</a></code>.</li>
</ol>
</div>
<div class="section level4">
<h4 id="wmean">
<code>wmean()</code><a class="anchor" aria-label="anchor" href="#wmean"></a>
</h4>
<p><code><a href="../reference/wmean.html">wmean()</a></code> creates weighted-mean traits, which represent
the most sophisticated yet powerful type of derived traits. These traits
calculate the abundance-weighted average of quantitative properties
across glycan populations. Examples include the average number of
antennae across all glycans, or the average degree of sialylation per
galactose across the entire glycan repertoire.</p>
<p>Before exploring <code><a href="../reference/wmean.html">wmean()</a></code> usage, it is essential to
understand the weighted-mean concept.</p>
<p>Consider three glycans: <code>G1</code>, <code>G2</code>, and
<code>G3</code>. Their respective antennae counts are 1, 2, and 3, with
relative abundances of 50%, 20%, and 30%. The weighted-mean number of
antennae is calculated as:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="fl">1</span> <span class="op">*</span> <span class="fl">0.5</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="fl">0.2</span> <span class="op">+</span> <span class="fl">3</span> <span class="op">*</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fl">0.5</span> <span class="op">+</span> <span class="fl">0.2</span> <span class="op">+</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1.8</span></span></code></pre></div>
<p>This value represents the abundance-weighted average degree of
branching across the glycan population.</p>
<p>This concept provides substantial analytical power, enabling the
calculation of numerous biologically meaningful traits.</p>
<p>Consider another example: the average degree of sialylation per
antenna across all glycans. Using the same three glycans
(<code>G1</code>, <code>G2</code>, and <code>G3</code>) with antennae
counts of 1, 2, and 3, sialic acid counts of 1, 1, and 3, respectively,
and relative abundances of 50%, 20%, and 30%.</p>
<p>First, calculate the sialylation degree per antenna for each glycan
(sialic acid count divided by antenna count):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1.0 0.5 1.0</span></span></code></pre></div>
<p>Subsequently, incorporate the abundance weighting:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">1</span> <span class="op">*</span> <span class="fl">0.5</span> <span class="op">+</span> <span class="fl">1</span><span class="op">/</span><span class="fl">2</span> <span class="op">*</span> <span class="fl">0.2</span> <span class="op">+</span> <span class="fl">3</span><span class="op">/</span><span class="fl">3</span> <span class="op">*</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fl">0.5</span> <span class="op">+</span> <span class="fl">0.2</span> <span class="op">+</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.9</span></span></code></pre></div>
<p>This yields the abundance-weighted average degree of sialylation per
antenna across the glycan population.</p>
<p>Having established the weighted-mean concept and mastered
<code><a href="../reference/prop.html">prop()</a></code> and <code><a href="../reference/ratio.html">ratio()</a></code> fundamentals, we can now
examine <code><a href="../reference/wmean.html">wmean()</a></code> implementation.</p>
<p><code><a href="../reference/wmean.html">wmean()</a></code> accepts expressions that evaluate to numeric
vectors, contrasting with the logical vectors required by
<code><a href="../reference/prop.html">prop()</a></code> and <code><a href="../reference/ratio.html">ratio()</a></code>.</p>
<p>For example, the average number of antennae across all glycans is
defined as:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nA</span><span class="op">)</span></span></code></pre></div>
<p>The average degree of sialylation per antenna across all glycans is
defined as:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nS</span> <span class="op">/</span> <span class="va">nA</span><span class="op">)</span></span></code></pre></div>
<p>The <code>within</code> parameter can be utilized to restrict the
weighted-mean calculation to specific glycan subsets. For example, the
average number of antennae within complex glycans is defined as:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nA</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Practice exercises:</p>
<ol style="list-style-type: decimal">
<li>Define a trait calculating the average degree of sialylation per
antenna.</li>
<li>Define a trait calculating the average number of arm fucoses.</li>
</ol>
</div>
<div class="section level4">
<h4 id="total">
<code>total()</code><a class="anchor" aria-label="anchor" href="#total"></a>
</h4>
<p><code><a href="../reference/total.html">total()</a></code> creates total abundance traits, which is the sum
of the abundances of a group of glycans. This is the simplest type of
derived traits in <code>glydet</code>.</p>
<p>For example, the total abundance of all complex glycans is defined
as:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/total.html">total</a></span><span class="op">(</span><span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span><span class="op">)</span></span></code></pre></div>
<p>There is no <code>within</code> parameter for <code><a href="../reference/total.html">total()</a></code>,
as you can always add a restriction to the expression by using
<code>&amp;</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/total.html">total</a></span><span class="op">(</span><span class="va">Tp</span> <span class="op">==</span> <span class="st">"complex"</span> <span class="op">&amp;</span> <span class="va">nA</span> <span class="op">==</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="wsum">
<code>wsum()</code><a class="anchor" aria-label="anchor" href="#wsum"></a>
</h4>
<p><code><a href="../reference/wsum.html">wsum()</a></code> is the cousin of <code><a href="../reference/wmean.html">wmean()</a></code>, but
instead of calculating the average value, it calculates the sum of the
values.</p>
<p>Let’s see an example to understand the difference between
<code><a href="../reference/wmean.html">wmean()</a></code> and <code><a href="../reference/wsum.html">wsum()</a></code>. Say we have three glycans:
<code>G1</code>, <code>G2</code>, and <code>G3</code> in two samples:
<code>S1</code> and <code>S2</code>.</p>
<ul>
<li>Sample S1: <code>G1</code> (10), <code>G2</code> (20),
<code>G3</code> (30) (intensity values)</li>
<li>Sample S2: <code>G1</code> (20), <code>G2</code> (40),
<code>G3</code> (60) (intensity values)</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expr_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">20</span>, <span class="fl">40</span>, <span class="fl">60</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expr_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G1"</span>, <span class="st">"G2"</span>, <span class="st">"G3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">expr_mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S1"</span>, <span class="st">"S2"</span><span class="op">)</span></span>
<span><span class="va">expr_mat</span></span>
<span><span class="co">#&gt;    S1 S2</span></span>
<span><span class="co">#&gt; G1 10 20</span></span>
<span><span class="co">#&gt; G2 20 40</span></span>
<span><span class="co">#&gt; G3 30 60</span></span></code></pre></div>
<p>For simplicity, we don’t use any meta-properties here, just the
number 1.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trait1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">trait1</span><span class="op">(</span><span class="va">expr_mat</span>, <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1 1</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trait2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wsum.html">wsum</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">trait2</span><span class="op">(</span><span class="va">expr_mat</span>, <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  60 120</span></span></code></pre></div>
<p>See the difference? <code><a href="../reference/wmean.html">wmean()</a></code> calculates the average value
of some numeric property, so as long as the ratio of each glycan’s
abundance is the same, the result is always the same. But
<code><a href="../reference/wsum.html">wsum()</a></code> is sensitive to the abundance of each glycan, so the
result is different.</p>
</div>
</div>
<div class="section level3">
<h3 id="using-custom-traits">Using Custom Traits<a class="anchor" aria-label="anchor" href="#using-custom-traits"></a>
</h3>
<p>Having learned how to define custom traits, we can now implement
them.</p>
<p><code><a href="../reference/derive_traits.html">derive_traits()</a></code> includes a <code>trait_fns</code>
parameter that accepts a named list of derived trait functions. Note
that the three factory functions described above return functions as
their output.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="fu"><a href="../reference/prop.html">prop</a></span><span class="op">(</span><span class="va">nFc</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "glydet_prop"  "glydet_trait"</span></span></code></pre></div>
<p>R supports functional programming paradigms, treating functions as
first-class objects. Functions can be passed as arguments to other
functions and returned from function calls.</p>
<p>While understanding this concept enhances R proficiency, it is not
prerequisite for usage—simply define traits as a named list and provide
it to the <code>trait_fns</code> parameter.</p>
<p>The following example examines sialylation degree within glycan
subsets of varying antenna counts:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  A2S <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nS</span> <span class="op">/</span> <span class="va">nA</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  A3S <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nS</span> <span class="op">/</span> <span class="va">nA</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  A4S <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nS</span> <span class="op">/</span> <span class="va">nA</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp</span>, trait_fns <span class="op">=</span> <span class="va">my_traits</span><span class="op">)</span></span></code></pre></div>
<p>The identifiers “A2S”, “A3S”, and “A4S” represent the derived trait
names.</p>
<p>When <code>trait_fns</code> is omitted, <code><a href="../reference/derive_traits.html">derive_traits()</a></code>
internally invokes <code><a href="../reference/basic_traits.html">basic_traits()</a></code> to utilize built-in trait
functions.</p>
</div>
<div class="section level3">
<h3 id="validating-trait-definitions">Validating Trait Definitions<a class="anchor" aria-label="anchor" href="#validating-trait-definitions"></a>
</h3>
<p>How can you ensure your trait definitions are accurate and
meaningful? The <code><a href="../reference/explain_trait.html">explain_trait()</a></code> function provides an
intuitive way to validate your trait definitions by generating
human-readable explanations.</p>
<p>This function helps you:</p>
<ul>
<li>Verify that your trait logic matches your intended analysis</li>
<li>Understand complex trait expressions at a glance<br>
</li>
<li>Debug and refine trait definitions before deployment</li>
</ul>
<p>Let’s examine the <code>A2S</code> trait defined earlier as an
example:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/explain_trait.html">explain_trait</a></span><span class="op">(</span><span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nS</span> <span class="op">/</span> <span class="va">nA</span>, within <span class="op">=</span> <span class="op">(</span><span class="va">nA</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Abundance-weighted mean of degree of sialylation per antenna within bi-antennary glycans."</span></span></code></pre></div>
<p>The function interprets your trait expression and returns a clear,
natural language description of what the trait calculates. This
validation step is particularly valuable when working with complex trait
definitions or when collaborating with team members who need to
understand your analytical approach.</p>
<p>We have introduced a new argument <code>use_ai</code> to
<code><a href="../reference/explain_trait.html">explain_trait()</a></code> since glydet 0.7.0. If <code>use_ai</code>
is TRUE, the function will use a Large Language Model (LLM) to explain
the trait, which covers more complex cases and edge cases. Check out the
documentation of <code><a href="../reference/explain_trait.html">explain_trait()</a></code> for details.</p>
</div>
</div>
<div class="section level2">
<h2 id="custom-meta-properties">Custom Meta-Properties<a class="anchor" aria-label="anchor" href="#custom-meta-properties"></a>
</h2>
<p>The preceding examples utilized exclusively built-in meta-properties.
As the building blocks of derived traits, understanding how to define
custom meta-properties will bring you more flexibility.</p>
<p><code>glydet</code> provides two means of defining custom
meta-properties:</p>
<ul>
<li>By providing meta-property functions to the <code>mp_fns</code>
parameter of <code><a href="../reference/derive_traits.html">derive_traits()</a></code>.</li>
<li>By using columns in the variable information tibble as
meta-properties.</li>
</ul>
<div class="section level3">
<h3 id="defining-custom-meta-properties-with-functions">Defining Custom Meta-Properties with Functions<a class="anchor" aria-label="anchor" href="#defining-custom-meta-properties-with-functions"></a>
</h3>
<p>Meta-property functions must accept a
<code><a href="https://glycoverse.github.io/glyrepr/reference/glycan_structure.html" class="external-link">glyrepr::glycan_structure()</a></code> vector and return a vector of
corresponding meta-property values.</p>
<p>For example, the built-in meta-property <code>B</code> is defined as
follows:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simplified version for illustration purposes</span></span>
<span><span class="kw">function</span><span class="op">(</span><span class="va">glycans</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">motif</span> <span class="op">&lt;-</span> <span class="st">"HexNAc(??-?)Hex(??-?)HexNAc(??-?)HexNAc(??-"</span></span>
<span>  <span class="fu">glymotif</span><span class="fu">::</span><span class="fu"><a href="https://glycoverse.github.io/glymotif/reference/have_motif.html" class="external-link">have_motif</a></span><span class="op">(</span><span class="va">glycans</span>, <span class="va">motif</span>, alignment <span class="op">=</span> <span class="st">"core"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>This implementation utilizes <code><a href="https://glycoverse.github.io/glymotif/reference/have_motif.html" class="external-link">glymotif::have_motif()</a></code> to
determine motif presence within glycan structures. Meta-property
function definitions offer considerable flexibility. Any tools from the
<code>glycoverse</code> ecosystem or broader R environment can be
employed, provided the function accepts a
<code><a href="https://glycoverse.github.io/glyrepr/reference/glycan_structure.html" class="external-link">glyrepr::glycan_structure()</a></code> vector and returns an atomic
vector. All built-in meta-properties utilize the <code>glymotif</code>
package for motif matching and counting operations. For detailed
information, consult the <a href="https://glycoverse.github.io/glymotif/" class="external-link">glymotif
documentation</a>.</p>
<p>The following example demonstrates custom meta-property creation:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_mp_fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="co"># Number of Lewis antigens</span></span>
<span>  nLe <span class="op">=</span> <span class="op">~</span> <span class="fu">glymotif</span><span class="fu">::</span><span class="fu"><a href="https://glycoverse.github.io/glymotif/reference/count_motif.html" class="external-link">count_motif</a></span><span class="op">(</span><span class="va">.x</span>, <span class="st">"Hex(??-?)[dHex(??-?)]HexNAc(??-"</span><span class="op">)</span>,</span>
<span>  <span class="co"># Number of poly-LacNAc units</span></span>
<span>  nPl <span class="op">=</span> <span class="op">~</span> <span class="fu">glymotif</span><span class="fu">::</span><span class="fu"><a href="https://glycoverse.github.io/glymotif/reference/count_motif.html" class="external-link">count_motif</a></span><span class="op">(</span><span class="va">.x</span>, <span class="st">"Hex(??-?)HexNAc(??-?)Hex(??-?)HexNAc(??-"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Note that purrr-style lambda functions provide a concise syntax for
meta-property function definitions.</p>
<p>Traits can be defined using these new meta-properties:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="co"># Average number of Lewis antigens per antenna</span></span>
<span>  LeA <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nLe</span> <span class="op">/</span> <span class="va">nA</span><span class="op">)</span>,</span>
<span>  <span class="co"># Average number of poly-LacNAc units per glycan</span></span>
<span>  Pl <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nPl</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>These expressions incorporate both custom meta-properties
(<code>nLe</code> and <code>nPl</code>) and the built-in meta-property
<code>nA</code>.</p>
<p>Implementation of these custom meta-properties and traits proceeds as
follows:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp</span>, trait_fns <span class="op">=</span> <span class="va">my_traits</span>, mp_fns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">my_mp_fns</span>, <span class="fu"><a href="../reference/all_mp_fns.html">all_mp_fns</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Traitproteomics Experiment</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Expression matrix: <span style="color: #0000BB;">12</span> samples, <span style="color: #0000BB;">552</span> variables</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Sample information fields: <span style="color: #00BB00;">group</span> <span style="color: #0000BB;">&lt;fct&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Variable information fields: <span style="color: #00BB00;">protein</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">protein_site</span> <span style="color: #0000BB;">&lt;int&gt;</span>, <span style="color: #00BB00;">trait</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">gene</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">explanation</span> <span style="color: #0000BB;">&lt;chr&gt;</span></span></span></code></pre></div>
<p>Ensure that custom meta-properties are combined with built-in
meta-properties when the latter are required within trait
expressions.</p>
<p>Let’s see another example about sialic acid linkage types. Neu5Ac can
have two different linkage types: a2-3 and a2-6. We can define two
custom meta-properties to count the number of sialic acids with each
linkage type:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sia_mp_fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="co"># Number of a2-3 sialic acids</span></span>
<span>  nL <span class="op">=</span> <span class="op">~</span> <span class="fu">glymotif</span><span class="fu">::</span><span class="fu"><a href="https://glycoverse.github.io/glymotif/reference/count_motif.html" class="external-link">count_motif</a></span><span class="op">(</span><span class="va">.x</span>, <span class="st">"NeuAc(a2-3)Hex(??-"</span><span class="op">)</span>,</span>
<span>  <span class="co"># Number of a2-6 sialic acids</span></span>
<span>  nE <span class="op">=</span> <span class="op">~</span> <span class="fu">glymotif</span><span class="fu">::</span><span class="fu"><a href="https://glycoverse.github.io/glymotif/reference/count_motif.html" class="external-link">count_motif</a></span><span class="op">(</span><span class="va">.x</span>, <span class="st">"NeuAc(a2-6)Hex(??-"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And define two traits to calculate the degree of a2-3 and a2-6
sialylation per galactose:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sia_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="co"># Average degree of a2-3 sialylation per galactose</span></span>
<span>  GL <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nL</span> <span class="op">/</span> <span class="va">nG</span><span class="op">)</span>,</span>
<span>  <span class="co"># Average degree of a2-6 sialylation per galactose</span></span>
<span>  GE <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nE</span> <span class="op">/</span> <span class="va">nG</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And calculate the traits:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp</span>, trait_fns <span class="op">=</span> <span class="va">sia_traits</span>, mp_fns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sia_mp_fns</span>, <span class="fu"><a href="../reference/all_mp_fns.html">all_mp_fns</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Traitproteomics Experiment</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Expression matrix: <span style="color: #0000BB;">12</span> samples, <span style="color: #0000BB;">552</span> variables</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Sample information fields: <span style="color: #00BB00;">group</span> <span style="color: #0000BB;">&lt;fct&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Variable information fields: <span style="color: #00BB00;">protein</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">protein_site</span> <span style="color: #0000BB;">&lt;int&gt;</span>, <span style="color: #00BB00;">trait</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">gene</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">explanation</span> <span style="color: #0000BB;">&lt;chr&gt;</span></span></span></code></pre></div>
<p>This is an example of how you can violate the <a href="https://glycoverse.github.io/glydet/articles/glydet.html#working-with-structural-ambiguity">ambiguity
assumption</a> of built-in meta-properties and derived traits, by
introducing more sophisticated and linkage-awared meta-properties.</p>
</div>
<div class="section level3">
<h3 id="defining-custom-meta-properties-with-columns">Defining Custom Meta-Properties with Columns<a class="anchor" aria-label="anchor" href="#defining-custom-meta-properties-with-columns"></a>
</h3>
<p>The first method is handy if all information you need is already
encoded in glycan structures. However, there are situations that you
need to rely on other meta-data. In this case, you can use columns in
the variable information tibble as meta-properties directly, by
specifying the <code>mp_cols</code> parameter.</p>
<p>For example, let’s use the sialic acid linkage type example above.
You might have used special derivatization methods to differentiate the
two linkage types. This information might end up in a column in the
variable information tibble, not directly in the glycan structures.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here we assume all sialic acids are a2-6</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="va">exp</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/mutate_obs.html" class="external-link">mutate_var</a></span><span class="op">(</span></span>
<span>    n_a26_sia <span class="op">=</span> <span class="fu"><a href="https://glycoverse.github.io/glyrepr/reference/count_mono.html" class="external-link">count_mono</a></span><span class="op">(</span><span class="va">glycan_structure</span>, <span class="st">"NeuAc"</span><span class="op">)</span>,</span>
<span>    n_a23_sia <span class="op">=</span> <span class="fl">0L</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp2</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n_a26_sia</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">glycan_structure</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;glycan_structure[2575]&gt;</span></span>
<span><span class="co">#&gt; [1] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [2] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)[<span style="color: #000000;">HexNAc</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [3] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [4] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [5] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [6] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">dHex</span>(??-?)]<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [7] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">dHex</span>(??-?)]<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [8] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)[<span style="color: #000000;">dHex</span>(??-?)]<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [9] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)[<span style="color: #000000;">dHex</span>(??-?)]<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; [10] <span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)[<span style="color: #000000;">NeuAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">Hex</span>(??-?)]<span style="color: #000000;">Hex</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-?)<span style="color: #000000;">HexNAc</span>(??-</span></span>
<span><span class="co">#&gt; ... (2565 more not shown)</span></span>
<span><span class="co">#&gt; # Unique structures: 535</span></span></code></pre></div>
<p>See? No linkage information about sialic acids is in the glycan
structures. But we have two columns: <code>n_a26_sia</code> and
<code>n_a23_sia</code> storing this information.</p>
<p>Now we can define two traits to calculate the degree of a2-3 and a2-6
sialylation per galactose:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sia_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="co"># Average degree of a2-3 sialylation per galactose</span></span>
<span>  GL <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nL</span> <span class="op">/</span> <span class="va">nG</span><span class="op">)</span>,</span>
<span>  <span class="co"># Average degree of a2-6 sialylation per galactose</span></span>
<span>  GE <span class="op">=</span> <span class="fu"><a href="../reference/wmean.html">wmean</a></span><span class="op">(</span><span class="va">nE</span> <span class="op">/</span> <span class="va">nG</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And use <code>mp_cols</code> to tell <code><a href="../reference/derive_traits.html">derive_traits()</a></code> to
use these columns as meta-properties:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp2</span>, trait_fns <span class="op">=</span> <span class="va">sia_traits</span>, mp_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>nL <span class="op">=</span> <span class="st">"n_a23_sia"</span>, nE <span class="op">=</span> <span class="st">"n_a26_sia"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Traitproteomics Experiment</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Expression matrix: <span style="color: #0000BB;">12</span> samples, <span style="color: #0000BB;">552</span> variables</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Sample information fields: <span style="color: #00BB00;">group</span> <span style="color: #0000BB;">&lt;fct&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Variable information fields: <span style="color: #00BB00;">protein</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">protein_site</span> <span style="color: #0000BB;">&lt;int&gt;</span>, <span style="color: #00BB00;">trait</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">gene</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">n_a23_sia</span> <span style="color: #0000BB;">&lt;int&gt;</span>, <span style="color: #00BB00;">explanation</span> <span style="color: #0000BB;">&lt;chr&gt;</span></span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="using-make_trait">Using <code>make_trait()</code><a class="anchor" aria-label="anchor" href="#using-make_trait"></a>
</h2>
<p><code><a href="../reference/make_trait.html">make_trait()</a></code> allows you to create a trait from natural
language. To use this feature, you need to install the
<code>ellmer</code> package (not a forced dependency of
<code>glydet</code>). You also need to provide an API key for the
DeepSeek chat model. You can obtain an API key from <a href="https://platform.deepseek.com" class="external-link uri">https://platform.deepseek.com</a>. Please set the
environment variable <code>DEEPSEEK_API_KEY</code> to your API key using
<code><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv()</a></code>:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Works for a single session</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>DEEPSEEK_API_KEY <span class="op">=</span> <span class="st">"your_api_key"</span><span class="op">)</span></span></code></pre></div>
<p>Then you can transform your ideas into trait functions by calling
<code><a href="../reference/make_trait.html">make_trait()</a></code>:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  nS <span class="op">=</span> <span class="fu"><a href="../reference/make_trait.html">make_trait</a></span><span class="op">(</span><span class="st">"the average number of sialic acids"</span><span class="op">)</span>,</span>
<span>  nG <span class="op">=</span> <span class="fu"><a href="../reference/make_trait.html">make_trait</a></span><span class="op">(</span><span class="st">"the average number of galactoses"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp</span>, trait_fns <span class="op">=</span> <span class="va">my_traits</span><span class="op">)</span></span></code></pre></div>
<p>This function is experimental, so the result should be verified
manually. If the description is not clear, an error will be raised. Try
to read the descriptions of built-in traits to get ideas.</p>
</div>
<div class="section level2">
<h2 id="exercise-solutions">Exercise Solutions<a class="anchor" aria-label="anchor" href="#exercise-solutions"></a>
</h2>
<p><strong><code><a href="../reference/prop.html">prop()</a></code></strong></p>
<ol style="list-style-type: decimal">
<li><code>prop(nGt &gt; 0)</code></li>
<li><code>prop(nGt &gt; 0 &amp; nS == 0)</code></li>
<li><code>prop(nA == 2 &amp; !B)</code></li>
<li><code>prop(B, within = (nA == 2))</code></li>
</ol>
<p><strong><code><a href="../reference/ratio.html">ratio()</a></code></strong></p>
<ol style="list-style-type: decimal">
<li><code>ratio(nFc &gt; 0, nFc == 0)</code></li>
<li><code>ratio(nA == 4, nA == 3, within = (Tp == "complex"))</code></li>
<li><code>ratio(nA == 2 &amp; Tp == "complex", Tp == "complex")</code></li>
</ol>
<p><strong><code><a href="../reference/wmean.html">wmean()</a></code></strong></p>
<ol style="list-style-type: decimal">
<li><code>wmean(nS / nA)</code></li>
<li><code>wmean(nFa)</code></li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
