<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quantifying Glycan Motifs • glydet</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quantifying Glycan Motifs">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">glydet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.3.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/glydet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/custom-traits.html">Defining Custom Traits</a></li>
    <li><a class="dropdown-item" href="../articles/quantify-motifs.html">Quantifying Glycan Motifs</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Quantifying Glycan Motifs</h1>
            
      

      <div class="d-none name"><code>quantify-motifs.Rmd</code></div>
    </div>

    
    
<p>In this vignette, we’ll explore a special type of derived trait:
motif quantification. This powerful feature allows us to measure the
abundance of biologically meaningful glycan substructures across your
samples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glydet/">glydet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyexp/" class="external-link">glyexp</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glycoverse.github.io/glyclean/" class="external-link">glyclean</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'glyclean'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     aggregate</span></span>
<span></span>
<span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://glycoverse.github.io/glyclean/reference/auto_clean.html" class="external-link">auto_clean</a></span><span class="op">(</span><span class="va">real_experiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Normalizing data (Median)</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Normalizing data (Median) <span style="color: #B2B2B2;">[137ms]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Removing variables with &gt;50% missing values</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Removing variables with &gt;50% missing values <span style="color: #B2B2B2;">[70ms]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Imputing missing values</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Sample size &lt;= 30, using sample minimum imputation</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Imputing missing values<span style="color: #00BB00;">✔</span> Imputing missing values <span style="color: #B2B2B2;">[23ms]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Aggregating data</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Aggregating data <span style="color: #B2B2B2;">[1.1s]</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Normalizing data again</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Normalizing data again <span style="color: #B2B2B2;">[18ms]</span></span></span></code></pre></div>
<div class="section level2">
<h2 id="what-is-motif-quantification">What is motif quantification?<a class="anchor" aria-label="anchor" href="#what-is-motif-quantification"></a>
</h2>
<p>Glycan motifs are substructures with special biological or structural
significance. Think of them as functional building blocks that carry
specific biological messages. For example, the Lewis x antigen is a
fucosylated carbohydrate epitope commonly found on glycoproteins and
glycolipids. It plays crucial roles in cell–cell recognition, adhesion,
and immune responses. Understanding the abundance of Lewis x antigens
can provide valuable insights into immune system dynamics.</p>
<p>But here’s the challenge: how do we quantify Lewis x antigens across
samples? Different glycans contain varying numbers of Lewis x motifs,
and these glycans themselves have different abundances across samples.
Our solution elegantly combines both pieces of information to estimate
the true abundance of Lewis x antigens in your dataset.</p>
</div>
<div class="section level2">
<h2 id="the-quantify_motifs-function">The <code>quantify_motifs()</code> function<a class="anchor" aria-label="anchor" href="#the-quantify_motifs-function"></a>
</h2>
<p>Glydet provides the <code><a href="../reference/quantify_motifs.html">quantify_motifs()</a></code> function to handle
this complex task seamlessly. This function takes a
<code><a href="https://glycoverse.github.io/glyexp/reference/experiment.html" class="external-link">glyexp::experiment()</a></code> object along with your motifs of
interest and returns a new experiment object enriched with motif
quantifications. Just like <code><a href="../reference/derive_traits.html">derive_traits()</a></code>,
<code><a href="../reference/quantify_motifs.html">quantify_motifs()</a></code> works beautifully with both glycomics and
glycoproteomics data.</p>
<p>Let’s dive into a practical example:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define our motifs of interest using IUPAC-condensed format</span></span>
<span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  Lx <span class="op">=</span> <span class="st">"Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>,  <span class="co"># Lewis x antigen</span></span>
<span>  SLx <span class="op">=</span> <span class="st">"NeuAc(??-?)Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>  <span class="co"># Sialyl Lewis x antigen</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quantify the motifs in our dataset</span></span>
<span><span class="va">motif_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify_motifs.html">quantify_motifs</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span><span class="op">)</span></span>
<span><span class="va">motif_exp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Traitproteomics Experiment</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Expression matrix: <span style="color: #0000BB;">12</span> samples, <span style="color: #0000BB;">548</span> variables</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Sample information fields: <span style="color: #00BB00;">group</span> <span style="color: #0000BB;">&lt;fct&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Variable information fields: <span style="color: #00BB00;">protein</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">protein_site</span> <span style="color: #0000BB;">&lt;int&gt;</span>, <span style="color: #00BB00;">motif</span> <span style="color: #0000BB;">&lt;chr&gt;</span>, <span style="color: #00BB00;">gene</span> <span style="color: #0000BB;">&lt;chr&gt;</span></span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://glycoverse.github.io/glyexp/reference/get_var_info.html" class="external-link">get_var_info</a></span><span class="op">(</span><span class="va">motif_exp</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 548 × 5</span></span></span>
<span><span class="co">#&gt;    variable protein protein_site motif gene  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> V1       A6NJW9            49 Lx    CD8B2 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> V2       A6NJW9            49 SLx   CD8B2 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> V3       O14786           150 Lx    NRP1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> V4       O14786           150 SLx   NRP1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> V5       O43866           226 Lx    CD5L  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> V6       O43866           226 SLx   CD5L  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> V7       O75437           244 Lx    ZNF254</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> V8       O75437           244 SLx   ZNF254</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> V9       O75581           281 Lx    LRP6  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> V10      O75581           281 SLx   LRP6  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 538 more rows</span></span></span></code></pre></div>
<p>Notice how the variable information now includes a <code>motif</code>
column instead of the usual <code>trait</code> column. Don’t worry—this
is just a cosmetic difference. Under the hood, the functionality works
exactly the same way as <code><a href="../reference/derive_traits.html">derive_traits()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="absolute-vs--relative-motif-quantification">Absolute vs. relative motif quantification<a class="anchor" aria-label="anchor" href="#absolute-vs--relative-motif-quantification"></a>
</h2>
<p>Here’s where things get a little bit complex: motif quantification
can be performed in two distinct ways—absolute and relative. By default,
<code><a href="../reference/quantify_motifs.html">quantify_motifs()</a></code> uses relative quantification, but
understanding both approaches is crucial for choosing the right method
for your research.</p>
<p><strong>Important note:</strong> Don’t confuse this with the absolute
and relative quantification of glycans or glycopeptides themselves—
that’s a different concept entirely.</p>
<p>In traditional omics contexts, “absolute quantification” means
determining actual concentrations (like mg/L) or counts (like mRNA copy
numbers), while “relative quantification” means comparing abundance
between samples without meaningful absolute values. Label-free
quantification, for instance, is a relative method.</p>
<p>However, in motif quantification, “absolute” and “relative” refer to
whether we normalize motif abundances by the total abundance of the
glycome (or individual glycosite mini-glycomes).</p>
<p>Let’s illustrate this with a concrete example. Imagine a glycomics
dataset with two samples, A and B, each containing three glycans (G1,
G2, G3) with these abundances:</p>
<ul>
<li>
<strong>Sample A:</strong> G1 = 10, G2 = 20, G3 = 30</li>
<li>
<strong>Sample B:</strong> G1 = 20, G2 = 40, G3 = 60</li>
</ul>
<p>For simplicity, let’s assume our target motif appears exactly once in
each glycan.</p>
<div class="section level3">
<h3 id="absolute-motif-quantification">Absolute motif quantification<a class="anchor" aria-label="anchor" href="#absolute-motif-quantification"></a>
</h3>
<p>We simply sum the motif abundances across all glycans:</p>
<ul>
<li>
<strong>Sample A:</strong> 10 × 1 + 20 × 1 + 30 × 1 = 60</li>
<li>
<strong>Sample B:</strong> 20 × 1 + 40 × 1 + 60 × 1 = 120</li>
</ul>
<p>The results clearly differ between samples.</p>
</div>
<div class="section level3">
<h3 id="relative-motif-quantification">Relative motif quantification<a class="anchor" aria-label="anchor" href="#relative-motif-quantification"></a>
</h3>
<p>We normalize by the total glycan abundance:</p>
<ul>
<li>
<strong>Sample A:</strong> (10 × 1 + 20 × 1 + 30 × 1) ÷ (10 + 20 +
30) = 60 ÷ 60 = 1</li>
<li>
<strong>Sample B:</strong> (20 × 1 + 40 × 1 + 60 × 1) ÷ (20 + 40 +
60) = 120 ÷ 120 = 1</li>
</ul>
<p>The results are identical in two samples!</p>
<p>This distinction matters because these methods answer different
biological questions:</p>
<ul>
<li>
<strong>Absolute quantification</strong> asks: “How many motifs are
present in this sample?”</li>
<li>
<strong>Relative quantification</strong> asks: “If I randomly select
one glycan molecule, how many motifs would I expect to find on it in
average?”</li>
</ul>
</div>
<div class="section level3">
<h3 id="choosing-the-right-approach">Choosing the right approach<a class="anchor" aria-label="anchor" href="#choosing-the-right-approach"></a>
</h3>
<p>So which method should you use? The answer depends on both your data
type and research objectives. Here are some practical guidelines:</p>
<p><strong>For glycomics data:</strong> Use relative motif
quantification in most cases, since glycomics data is inherently
compositional (look up “compositional data analysis” if you’re curious
about the statistical theory behind this).</p>
<p><strong>For glycoproteomics data:</strong></p>
<ul>
<li>
<strong>Choose absolute quantification</strong> when you want to
compare observed motif abundance across samples. Keep in mind that many
factors can cause differences between samples, including upregulation of
enzymes responsible for the motif or changes in overall glycosylation
site occupancy.</li>
<li>
<strong>Choose relative quantification</strong> when you want to
understand underlying regulatory mechanisms by correcting for
differences in site occupancy. This approach helps isolate the
biological signal from technical variation.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="connection-to-derived-traits">Connection to derived traits<a class="anchor" aria-label="anchor" href="#connection-to-derived-traits"></a>
</h2>
<p>You might wonder why <code><a href="../reference/quantify_motifs.html">quantify_motifs()</a></code> lives in the
<code>glydet</code> package rather than <code>glymotif</code>. There’s
actually an interesting history here! Before <code>glymotif</code>
v0.9.0, there was indeed a <code><a href="../reference/quantify_motifs.html">quantify_motifs()</a></code> function in
<code>glymotif</code>. However, we eventually realized that motif
quantification is fundamentally a special case of derived traits, so we
reimplemented it in <code>glydet</code> with a more consistent and
powerful interface.</p>
<p>To demonstrate this connection, let’s manually perform motif
quantification using <code><a href="../reference/derive_traits.html">derive_traits()</a></code>. This will help you
understand that motif quantification is essentially just a specialized
application of trait derivation: <code><a href="../reference/wsum.html">wsum()</a></code> traits for absolute
quantification and <code><a href="../reference/wmean.html">wmean()</a></code> traits for relative
quantification.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, add the meta-properties to the variable information</span></span>
<span><span class="va">motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  nLx <span class="op">=</span> <span class="st">"Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>,  <span class="co"># Lewis x antigen</span></span>
<span>  nSLx <span class="op">=</span> <span class="st">"NeuAc(??-?)Hex(??-?)[dHex(??-?)]HexNAc(??-"</span>  <span class="co"># Sialyl Lewis x antigen</span></span>
<span><span class="op">)</span></span>
<span><span class="va">exp_with_mps</span> <span class="op">&lt;-</span> <span class="fu">glymotif</span><span class="fu">::</span><span class="fu"><a href="https://glycoverse.github.io/glymotif/reference/add_motifs_int.html" class="external-link">add_motifs_int</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define the traits using wsum() for absolute quantification</span></span>
<span><span class="va">trait_fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Lx <span class="op">=</span> <span class="fu"><a href="../reference/wsum.html">wsum</a></span><span class="op">(</span><span class="va">nLx</span><span class="op">)</span>, SLx <span class="op">=</span> <span class="fu"><a href="../reference/wsum.html">wsum</a></span><span class="op">(</span><span class="va">nSLx</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate the traits</span></span>
<span><span class="fu"><a href="../reference/derive_traits.html">derive_traits</a></span><span class="op">(</span><span class="va">exp_with_mps</span>, trait_fns <span class="op">=</span> <span class="va">trait_fns</span>, mp_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nLx"</span>, <span class="st">"nSLx"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This code snippet is functionally equivalent to:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The much simpler approach using quantify_motifs()</span></span>
<span><span class="fu"><a href="../reference/quantify_motifs.html">quantify_motifs</a></span><span class="op">(</span><span class="va">exp</span>, <span class="va">motifs</span>, method <span class="op">=</span> <span class="st">"absolute"</span><span class="op">)</span></span></code></pre></div>
<p>Pretty neat, right? The <code><a href="../reference/quantify_motifs.html">quantify_motifs()</a></code> function is
essentially doing all the heavy lifting shown in the manual approach,
but with additional robustness and user-friendly features.</p>
<p>For relative motif quantification, you’d simply replace
<code><a href="../reference/wsum.html">wsum()</a></code> with <code><a href="../reference/wmean.html">wmean()</a></code> in the manual approach,
while everything else stays the same. This flexibility showcases the
power of the underlying trait derivation system.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bin Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
